bolt-modules/boltlib/lib/puppet/functions/catch_errors.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Catches errors in a given block and returns them. This will return the output of the block if no errors are raised. Accepts an optional list of error kinds to catch.
+Puppet::Functions.create_function(:catch_errors) do
+  # @param error_types An array of error types to catch
+  # @param block The block of steps to catch errors on
+  # @return Undef If an error is raised in the block then the error will be returned, otherwise the result will be returned
+  # @example Catch errors for a block
+  # catch_errors() || {
+  #   run_command("whoami", $nodes)
+  #   run_task("foo::bar", $nodes)
+  # }
+  # @example Catch parse errors for a block of code
+  # catch_errors(['bolt/parse-error']) || {
+  #  run_plan('canary', $nodes)
+  #  run_plan('other_plan)
+  #  apply($nodes) || {
+  #    notify { "Hello": }
+  #  }
+  # }
+  dispatch :catch_errors do
+    optional_param 'Optional[Array[String[1]]]', :error_types
+    block_param 'Callable[0, 0]', :block
+  end
+
+  def catch_errors(error_types = nil)
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+      .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING,
+                            action: 'catch_errors')
+    end
+
+    # If it's a task that failed, that should be returned
+    # If it's a plan, plan failure, etc.
+    begin
+      yield
+    rescue => e
+      if e.is_a?(Puppet::PreformattedError)
```

Comment:
This is a bit different than the logic we use for `_catch_error` when running a plan, and I can't quite wrap my brain around how it ought to work.

What is the current return value (not including this function) of `run_plan('foo', _catch_errors => true)` if `foo` 
runs a task which errors? It seems like that should be the same as the return value of this function.